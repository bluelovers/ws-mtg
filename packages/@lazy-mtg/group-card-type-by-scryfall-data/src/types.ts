export const enum EnumGroupCardType
{
	Sorcery = 'Sorcery',
	Instant = 'Instant',
	Planeswalker = 'Planeswalker',
	Enchantment = 'Enchantment',
	Artifact = 'Artifact',
	Creature = 'Creature',
	Land = 'Land',

	Commander = 'Commander',
	Uncategorized = 'Uncategorized',

	Sideboard = 'Sideboard',
	Companion = 'Companion',
}

export const enum EnumRarity
{
	Common = 'common',
	Uncommon = 'uncommon',
	Rare = 'rare',
	Mythic = 'mythic'
}
