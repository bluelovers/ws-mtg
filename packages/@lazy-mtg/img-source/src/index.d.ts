import type { ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';
import { ICardForQueryInfo, ILinkURLConfig } from '@lazy-mtg/link-source';
export declare const enum EnumImgSource {
    gatherer = "gatherer",
    urzaCo = "urzaCo",
    scryfall = "scryfall",
    /**
     * @deprecated
     */
    magicCardsInfo = "magicCardsInfo"
}
export declare const enum EnumScryfallImageFormat {
    small = "small",
    normal = "normal",
    large = "large",
    png = "png",
    art_crop = "art_crop",
    border_crop = "border_crop"
}
export interface IImageURLConfig extends ILinkURLConfig {
    imgSource?: ITSTypeAndStringLiteral<EnumImgSource>;
    /**
     * only for scryfall
     */
    imageFormat?: ITSTypeAndStringLiteral<EnumScryfallImageFormat>;
    imageFaceBack?: boolean;
}
export declare function getImageSourceURL(card: ICardForQueryInfo, config?: IImageURLConfig): URL;
export default getImageSourceURL;
