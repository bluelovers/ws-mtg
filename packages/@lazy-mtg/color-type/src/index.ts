import { ITSArrayListMaybeReadonly, ITSValueOrArrayMaybeReadonly } from 'ts-type/lib/type/base';
import { ITSAndTypeAndStringLiteral } from 'ts-type/lib/helper/string';

export declare const enum EnumColor
{
	White = "W",
	Blue = "U",
	Black = "B",
	Red = "R",
	Green = "G"
}

export declare const enum EnumColorTitle
{
	W = "White",
	U = "Blue",
	B = "Black",
	R = "Red",
	G = "Green",

	Multi = 'Multi colored',
	Colorless = 'Colorless',
}

export const COLOR_PRIORITY = Object.freeze([
	EnumColor.White,
	EnumColor.Blue,
	EnumColor.Black,
	EnumColor.Red,
	EnumColor.Green,
] as const);

export function getColorTitle(color_identity: ITSValueOrArrayMaybeReadonly<ITSAndTypeAndStringLiteral<EnumColor>>)
{
	color_identity = [color_identity].flat() as EnumColor[];

	if (color_identity.length === 0)
	{
		return EnumColorTitle.Colorless
	}
	else if (color_identity.length === 1)
	{
		assertColorIdentity(color_identity[0]);

		// @ts-ignore
		return EnumColorTitle[color_identity[0]] as EnumColorTitle
	}

	return EnumColorTitle.Multi
}

export function isColorIdentity(color_identity: ITSAndTypeAndStringLiteral<EnumColor>): color_identity is EnumColor
{
	return COLOR_PRIORITY.includes(color_identity as EnumColor)
}

export function assertColorIdentity(color_identity: ITSAndTypeAndStringLiteral<EnumColor>): asserts color_identity is EnumColor
{
	if (!isColorIdentity(color_identity))
	{
		throw new TypeError(`Invalid Color Identity: ${color_identity}`)
	}
}

export function isColorIdentityArray(color_identity: ITSArrayListMaybeReadonly<ITSAndTypeAndStringLiteral<EnumColor>>): color_identity is ITSArrayListMaybeReadonly<EnumColor>
{
	return color_identity.every(isColorIdentity)
}

export function assertColorIdentityArray(color_identity: ITSArrayListMaybeReadonly<ITSAndTypeAndStringLiteral<EnumColor>>): asserts color_identity is ITSArrayListMaybeReadonly<EnumColor>
{
	color_identity.forEach(assertColorIdentity)
}

export function isColorIdentityLazy(color_identity: ITSValueOrArrayMaybeReadonly<ITSAndTypeAndStringLiteral<EnumColor>>)
{
	return isColorIdentityArray([color_identity].flat())
}

export function assertColorIdentityLazy(color_identity: ITSValueOrArrayMaybeReadonly<ITSAndTypeAndStringLiteral<EnumColor>>)
{
	return assertColorIdentityArray([color_identity].flat())
}

export default isColorIdentityLazy
