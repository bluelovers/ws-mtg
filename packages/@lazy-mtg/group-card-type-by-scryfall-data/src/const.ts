import { EnumGroupCardType } from './types';

export const GROUP_CARD_TYPE_PRIORITY = [
	EnumGroupCardType.Sorcery,
	EnumGroupCardType.Instant,
	EnumGroupCardType.Planeswalker,
	EnumGroupCardType.Enchantment,
	EnumGroupCardType.Artifact,
	EnumGroupCardType.Creature,
	EnumGroupCardType.Land,
] as const;

export const GROUP_CARD_TYPE_DISPLAY_PRIORITY = [
	EnumGroupCardType.Commander,
	EnumGroupCardType.Companion,

	EnumGroupCardType.Creature,
	EnumGroupCardType.Planeswalker,
	EnumGroupCardType.Artifact,
	EnumGroupCardType.Enchantment,
	EnumGroupCardType.Instant,
	EnumGroupCardType.Sorcery,
	EnumGroupCardType.Land,

	EnumGroupCardType.Sideboard,

	EnumGroupCardType.Uncategorized,
] as const;
