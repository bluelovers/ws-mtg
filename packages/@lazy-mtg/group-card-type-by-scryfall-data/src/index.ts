export { GROUP_CARD_TYPE_DISPLAY_PRIORITY } from './const';
export { GROUP_CARD_TYPE_PRIORITY } from './const';
export { EnumGroupCardType } from './types';
export { groupKeyByCardTypePriority } from './filter/card-type';
export * from './filter/_helper';
