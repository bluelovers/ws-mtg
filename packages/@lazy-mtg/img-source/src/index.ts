import type { ICardWithoutAmount } from 'mtg-decklist-parser2';
import type { ITSAndStringLiteral } from 'ts-type/lib/helper/string';

export const enum EnumImgSource
{
	gatherer = 'gatherer',
	urzaCo = 'urzaCo',
	scryfall = 'scryfall',
	/**
	 * @deprecated
	 */
	magicCardsInfo = 'magicCardsInfo',
}

function pad(num: string | number, size: number)
{
	return String(num).padStart(size, '0')
}

export interface ICardForQueryInfo extends Partial<Omit<ICardWithoutAmount, 'mtgoID'>>
{
	multiverseid?: string | number,
	mtgoID?: string | number,
}

export enum EnumScryfallImageFormat
{
	small = 'small',
	normal = 'normal',
	large = 'large',
	png = 'png',
	art_crop = 'art_crop',
	border_crop = 'border_crop',
}

export interface IImageURLConfig
{
	imgSource?: ITSAndStringLiteral<EnumImgSource>,
	language?: string,

	/**
	 * only for scryfall
	 */
	imageFormat?: ITSAndStringLiteral<EnumScryfallImageFormat>,
	imageFaceBack?: boolean,
}

function handleName(name: string)
{
	return name.split(/\s*\/\/\s*/)[0]
}

export function getImageSourceURL(card: ICardForQueryInfo, config?: IImageURLConfig)
{
	let imgUrl: URL | string;

	const cardNum = card.collectors;

	switch (config?.imgSource)
	{
		case EnumImgSource.magicCardsInfo:
		case EnumImgSource.scryfall:

			if (!cardNum)
			{
				if (card.multiverseid)
				{
					imgUrl = `https://api.scryfall.com/cards/multiverse/${card.multiverseid}?format=image`
				}
				else if (card.mtgoID)
				{
					imgUrl = new URL(`https://api.scryfall.com/cards/mtgo/${card.mtgoID}`);
				}
				else if (card.name?.length > 0)
				{
					imgUrl = new URL(`https://api.scryfall.com/cards/named`);

					imgUrl.searchParams.set('fuzzy', card.name);
				}
				else
				{
					throw new TypeError(`Invalid collector number: ${cardNum}`)
				}
			}
			else
			{
				const language = config?.language ?? 'en';

				imgUrl = `https://api.scryfall.com/cards/${card.set!.toLowerCase()}/${cardNum}/${language}`
			}

			if (typeof imgUrl === 'string')
			{
				imgUrl = new URL(imgUrl);
			}

			imgUrl.searchParams.set('format', 'image');

			if (config?.imageFormat)
			{
				imgUrl.searchParams.set('version', config.imageFormat);
			}

			if (config?.imageFaceBack)
			{
				imgUrl.searchParams.set('face', 'back');
			}

			break
		case EnumImgSource.urzaCo:

			if (!cardNum)
			{
				throw new TypeError(`Invalid collector number: ${cardNum}`)
			}

			let paddedNum = pad(cardNum, 3);
			// @ts-ignore
			if (isNaN(paddedNum.slice(-1)))
			{
				paddedNum = pad(paddedNum, 4);
			}

			imgUrl = `https://s3.urza.co/cards/${card.set!.toLowerCase()}/front/200dpi/${paddedNum}.jpg`;

			break
		default:
			if (!card!.multiverseid)
			{
				if (card.name?.length)
				{
					imgUrl = new URL(`https://gatherer.wizards.com/Handlers/Image.ashx?type=card`)

					imgUrl.searchParams.set('name', handleName(card.name));
				}
				else
				{
					throw new TypeError(`Invalid multiverseid: ${card!.multiverseid}`)
				}
			}
			else
			{
				imgUrl = `https://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=${card.multiverseid}&type=card`
			}

			break
	}

	if (typeof imgUrl === 'string')
	{
		imgUrl = new URL(imgUrl);
	}

	return imgUrl
}

export default getImageSourceURL
