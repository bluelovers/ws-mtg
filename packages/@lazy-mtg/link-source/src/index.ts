import type { ICardWithoutAmount } from 'mtg-decklist-parser2';
import type { ITSAndTypeAndStringLiteral, ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';
import escapeStringRegexp from 'escape-string-regexp';

export const enum EnumLinkSource
{
	gatherer = 'gatherer',
	urzaCo = 'urzaCo',
	scryfall = 'scryfall',
	/**
	 * @deprecated
	 */
	magicCardsInfo = 'magicCardsInfo',

	comboDeck = 'comboDeck',
}

export interface ICardForQueryInfo extends Partial<Omit<ICardWithoutAmount, 'mtgoID' | 'multiverseid'>>
{
	multiverseid?: string | number,
	mtgoID?: string | number,
}

export const enum EnumScryfallLanguages
{
	English = 'en',
	Japanese = 'ja',
	SimplifiedChinese = 'zhs',
	TraditionalChinese = 'zht',
	Phyrexian = 'ph',
}

export interface ILinkURLConfig
{
	linkSource?: ITSTypeAndStringLiteral<EnumLinkSource>,
	language?: ITSAndTypeAndStringLiteral<EnumScryfallLanguages>,
}

export function _handleName(name: string)
{
	return name.split(/\s*\/\/\s*/)[0]
}

export function _toURL(linkUrl: URL | string)
{
	if (typeof linkUrl === 'string')
	{
		linkUrl = new URL(linkUrl);
	}

	return linkUrl
}

export function _padZero(num: string | number, size: number)
{
	return String(num).padStart(size, '0')
}

export function getLinkSourceURL(card: ICardForQueryInfo, config?: ILinkURLConfig)
{
	let linkUrl: URL | string;

	switch (config?.linkSource)
	{
		case EnumLinkSource.magicCardsInfo:
		case EnumLinkSource.scryfall:

			linkUrl = `https://scryfall.com/search?q=`;

			if (card.multiverseid)
			{
				linkUrl += `multiverseid=${card.multiverseid}`
			}
			else if (card.mtgoID)
			{
				linkUrl += `mtgoID=${card.mtgoID}`
			}
			else if (card.set && card.collectors)
			{
				linkUrl += `set:${card.set.toLowerCase()} number:${card.collectors}`
			}
			else if (!card.name?.length)
			{
				throw new TypeError(`Invalid name: ${card.name}`)
			}
			else
			{
				linkUrl += `name:/(?<=\\/\\/ |^)${escapeStringRegexp(card.name).replace(/\//g, '\\/')}(?= \\/\\/|$)/`
			}

			if (config?.language)
			{
				linkUrl += ` lang:${config.language}`
			}

			break
		case EnumLinkSource.urzaCo:

			if (!card.name?.length)
			{
				throw new TypeError(`Invalid name: ${card.name}`)
			}

			linkUrl = new URL(`https://urza.co/m/cards`);
			linkUrl.searchParams.set('q', card.name);

			break
		case EnumLinkSource.comboDeck:

			if (!card.name?.length)
			{
				throw new TypeError(`Invalid name: ${card.name}`)
			}

			linkUrl = new URL(`http://combodeck.net/Query/${encodeURIComponent(card.name)}`);

			break
		default:
			if (!card!.multiverseid)
			{
				if (card.name?.length)
				{
					linkUrl = new URL(`https://gatherer.wizards.com/Pages/Card/Details.aspx?type=card`)

					linkUrl.searchParams.set('name', _handleName(card.name));
				}
				else
				{
					throw new TypeError(`Invalid multiverseid: ${card!.multiverseid}`)
				}
			}
			else
			{
				linkUrl = `https://gatherer.wizards.com/Pages/Card/Details.aspx?multiverseid=${card.multiverseid}&type=card`
			}

			break
	}

	return _toURL(linkUrl)
}

export default getLinkSourceURL
