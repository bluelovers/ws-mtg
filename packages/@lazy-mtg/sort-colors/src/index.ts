import { COLOR_PRIORITY, EnumColor } from '@lazy-mtg/color-type';
import { ITSAndTypeAndStringLiteral } from 'ts-type/lib/helper/string';

export { COLOR_PRIORITY, EnumColor }

/**
 * @see https://markrosewater.tumblr.com/post/13840097441/whats-the-proper-number-order-for-multicoloured
 */
export function compareWUBRG(a: ITSAndTypeAndStringLiteral<EnumColor>,
	b: ITSAndTypeAndStringLiteral<EnumColor>,
): 0 | 1 | -1
{
	const c1: EnumColor = a[0] as any;
	const c2: EnumColor = b[0] as any;

	const len1 = c1?.length;
	const len2 = c2?.length;

	if (!len1 || !len2)
	{
		if (len1)
		{
			return 1
		}
		else if (len2)
		{
			return -1
		}

		return 0
	}

	const v1 = COLOR_PRIORITY.indexOf(c1);
	const v2 = COLOR_PRIORITY.indexOf(c2);

	if (v1 === v2)
	{
		return compareWUBRG(a.slice(1), b.slice(1))
	}

	return v1 > v2 ? 1 : -1
}

export default compareWUBRG
