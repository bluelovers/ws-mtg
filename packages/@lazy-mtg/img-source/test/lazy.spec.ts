import { getImageSourceURL } from '../src/index';
import { EnumImgSource } from '../src/index';

const doubleFaceCard = 'Sea Gate Restoration // Sea Gate, Reborn';

describe(`gatherer`, () =>
{

	test(doubleFaceCard, () =>
	{

		let actual = getImageSourceURL({
			name: doubleFaceCard,
		});

		expect(actual).toMatchSnapshot();

	});

	test(`multiverseid=491706`, () =>
	{

		let actual = getImageSourceURL({
			multiverseid: 491706,
		});

		expect(actual).toMatchSnapshot();

	});

});

describe(`scryfall`, () =>
{

	test(doubleFaceCard, () =>
	{

		let actual = getImageSourceURL({
			name: doubleFaceCard,
		}, {
			imgSource: EnumImgSource.scryfall,
		});

		expect(actual).toMatchSnapshot();

	});

	test(`multiverseid=491706`, () =>
	{

		let actual = getImageSourceURL({
			multiverseid: 491706,
		}, {
			imgSource: EnumImgSource.scryfall,
		});

		expect(actual).toMatchSnapshot();

		actual = getImageSourceURL({
			multiverseid: 491706,
		}, {
			imgSource: EnumImgSource.scryfall,
			imageFaceBack: true,
		});

		expect(actual).toMatchSnapshot();

	});

	test(`mtgoID=83105`, () =>
	{

		let actual = getImageSourceURL({
			mtgoID: '83105',
		}, {
			imgSource: EnumImgSource.scryfall,
		});

		expect(actual).toMatchSnapshot();

		actual = getImageSourceURL({
			mtgoID: '83105',
		}, {
			imgSource: EnumImgSource.scryfall,
			imageFaceBack: true,
		});

		expect(actual).toMatchSnapshot();

	});

});
