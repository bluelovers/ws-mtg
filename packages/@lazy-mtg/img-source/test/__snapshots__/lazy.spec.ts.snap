// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatherer Sea Gate Restoration // Sea Gate, Reborn 1`] = `"https://gatherer.wizards.com/Handlers/Image.ashx?type=card&name=Sea+Gate+Restoration"`;

exports[`gatherer multiverseid=491706 1`] = `"https://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=491706&type=card"`;

exports[`scryfall Sea Gate Restoration // Sea Gate, Reborn 1`] = `"https://api.scryfall.com/cards/named?fuzzy=Sea+Gate+Restoration+%2F%2F+Sea+Gate%2C+Reborn&format=image"`;

exports[`scryfall mtgoID=83105 1`] = `"https://api.scryfall.com/cards/mtgo/83105?format=image"`;

exports[`scryfall mtgoID=83105 2`] = `"https://api.scryfall.com/cards/mtgo/83105?format=image&face=back"`;

exports[`scryfall multiverseid=491706 1`] = `"https://api.scryfall.com/cards/multiverse/491706?format=image"`;

exports[`scryfall multiverseid=491706 2`] = `"https://api.scryfall.com/cards/multiverse/491706?format=image&face=back"`;
